### ISSUE TITLE
internal IP not exposed to tasks w/ group bridge networking

### ISSUE THREAD

**Issue by tgross** [hashicorp/nomad#9781]
We don't expose the internal IP for a group bridged network to the tasks via an environment variable. The jobspec below has two groups, one with a group network and one with the deprecated task networking. The first group is also configured to advertise the internal IP to Consul. So while we're advertising the correct IP to Consul and we can reach the application on that IP address, if the application needs to know its own IP address (some applications use their IPs as unique IDs), we end up having to write some kind of prestart task to run `ip addr` to get it.

```hcl
job "example" {
  datacenters = ["dc1"]

  group "test1" {
    network {
      mode = "bridge"
      port "www" {
        to = 8001
      }
    }

    # internal IP + service port (ex 172.26.64.42:8001) gets advertised to Consul
    service {
      address_mode = "alloc"
      name         = "test1"
      port         = "8001"
    }

    task "server" {
      driver = "docker"
      config {
        image        = "busybox:1"
        command      = "httpd"
        args         = ["-v", "-f", "-p", "8001", "-h", "/var/www"]
        ports        = ["www"]
      }
    }
  }

  group "test2" {

    task "server" {
      driver = "docker"
      config {
        image        = "busybox:1"
        command      = "httpd"
        args         = ["-v", "-f", "-p", "8001", "-h", "/var/www"]
        port_map {
          www = 8001
        }
      }

      # host IP + mapped port (ex. 10.0.2.15:28870) gets advertised to Consul
      service {
        name         = "test2"
        port         = "www"
      }

      resources {
        network {
          port "www" {
            to = 8001
          }
        }
      }
    }
  }

}

```

The resulting environment for group 1, with group networking:

```sh
$ nomad alloc exec :alloc_id env | grep 'NOMAD.*IP\|PORT'
NOMAD_ALLOC_PORT_www=8001
NOMAD_HOST_IP_www=10.0.2.15
NOMAD_HOST_PORT_www=25274
NOMAD_IP_www=10.0.2.15
NOMAD_PORT_www=8001
```

And for group 2, with deprecated task networking:

```sh
$ nomad alloc exec c0a env | grep 'NOMAD.*IP\|PORT'
NOMAD_HOST_PORT_www=28870
NOMAD_IP_www=10.0.2.15
NOMAD_PORT_www=8001
```

cc @nickethier from today's internal discussion

**Comment by nickethier** [comment]
Brain dumping the context I have around this. The values we have today are all correct.

`NOMAD_IP_`, `NOMAD_PORT_` and `NOMAD_ADDR_` will exist as is to maintain backwards compatibility as shown above.

Now that we have the allocation IP available in the `AllocNetworkStatus` we need to plumb in two new network related fields for each port: `NOMAD_ALLOC_IP_` and `NOMAD_ALLOC_ADDR_`.

**Comment by tgross** [comment]
I did figure out a truly gross hack-around for this, which is that Consul will receive service registration before we start the task, so you can use a `template` to get the service address and port and write it to an environment variable: 

```hcl
job "example" {
  datacenters = ["dc1"]

  group "test" {
    network {
      mode = "bridge"
      port "www" {
        to = 8001
      }
    }

    service {
      address_mode = "alloc"
      name         = "mon"
      port         = "8001"
    }

    task "server" {
      driver = "docker"
      config {
        image   = "busybox:1"
        command = "httpd"
        args    = ["-v", "-f", "-p", "8001", "-h", "/var/www"]
        ports   = ["www"]
      }

      template {
        data = <<EOT
{{range $i, $services := service "mon"}}
MON_IP={{ .Address }}:{{ .Port }}
{{ end }}
EOT

        destination = "${NOMAD_TASK_DIR}/env"
        env         = true
      }
    }

  }

```

```
$ nomad alloc exec b90 env | grep MON_IP
MON_IP=172.26.64.78:8001
```

**Comment by tgross** [comment]
Just noting that this was a dupe of https://github.com/hashicorp/nomad/issues/8264 but we'll track the issue here.

**Comment by jrasell** [comment]
This work was conducted and merged via https://github.com/hashicorp/nomad/pull/11810 which was released as part of [1.8.0](https://github.com/hashicorp/nomad/blob/main/CHANGELOG.md#180-may-28-2024). Closing this issue as completed.
